### PR TITLE
Add GitHub authentication

### PR DESCRIPTION
**Pull request by vmx**
It's now possible to login with GitHub. The user won't need to create
a local account with username and password.

I tried to make this change as minimal and clean as possible. Though it's surely not ready to be merged, but I think it's good enough to get feedback on what is needed in order to get it merged.

## How Has This Been Tested?

Manually with logging in with a GitHub account. Also deleting the account again and logging in again. And that a lot of times during the development.

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My change is listed in the CHANGELOG.rst if appropriate.

### CODE REVIEW

**Comment by vmx** [comment]
It's great to see that you like the change. It was really meant as a starting point only to get third party authentication into the system eventually.

> adding one specific OAuth provider to pretalx seems very incongruent and random to me.

The plan surely is to have more providers supported. For keeping things simple, I started with one.
`django-allauth` uses the admin interface (which isn't enabled for pretalx) to setup other providers. It would also be possible to mimic that UI, so you could add any provider `django-allauth` supports. The advantage of having a custom interface for each "supported" provider is that you can leave out fields you don't need for that specific provider.

> make plugins like `pretalx-auth-github` or `pretalx-auth-twitter` etc a reality.

As those would all be handled via `django-allauth`, I'd rather do a `pretalx-auth-allauth` plugin instead which supports all providers. Or alternatively having `django-allauth` in the core and then having `pretalx-allauth-github` and `pretalx-allauth-twitter` plugins which just provide a nice customized UI for them.

>     * How do we handle /orga logins? They are not event specific, at the moment, while oauth configuration is per-event in the current draft.

To be honest, I didn't fully understand the distinction between `/orga` and the event specific stuff. It looked like use accounts are still central/unique to the whole system. So I just went ahead and made it work somehow. I think ideally you would configure one for the whole system and then enable different providers on an event basis.

**Comment by rixx** [comment]
> The advantage of having a custom interface for each "supported" provider is that you can leave out fields you don't need for that specific provider.

I'm all for custom-built interfaces, hence the idea of offering different plugins for the task.

> > make plugins like `pretalx-auth-github` or `pretalx-auth-twitter` etc a reality.
> 
> As those would all be handled via `django-allauth`, I'd rather do a `pretalx-auth-allauth` plugin instead which supports all providers. Or alternatively having `django-allauth` in the core and then having `pretalx-allauth-github` and `pretalx-allauth-twitter` plugins which just provide a nice customized UI for them.

Yes, these are the two reasonable options, and I think having one `pretalx-oauth` plugin would be easier to maintain, and it would house common code parts, so we should go with that.

> > * How do we handle /orga logins? They are not event specific, at the moment, while oauth configuration is per-event in the current draft.
> 
> To be honest, I didn't fully understand the distinction between `/orga` and the event specific stuff. It looked like use accounts are still central/unique to the whole system. So I just went ahead and made it work somehow. I think ideally you would configure one for the whole system and then enable different providers on an event basis.

Well, generally things work on a per-event basis in pretalx – events that have nothing in common can share an instance. So at least the choice of which providers to offer on login would be event specific, but the /orga login page is not namespaced under an event (one of the reasons being that you might not even have an event yet, and would still want to log in).

**Comment by vmx** [comment]
Thanks for the answers. What would the next steps be?

**Comment by vmx** [comment]
I should rephrase it to: What should I work on in order to make GitHub authentication in Pretalx possible :)

**Comment by rixx** [comment]
Sorry, yes, these last days were filled with stress and sickness, so I didn't get round to answering.

My proposal would be along these lines: I open a pretalx/pretalx-socialauth repository that will be the plugin for pretalx socialauth (of all kinds), with the goal of having GitHub at first (others as requested/required). You could work on that repository and figure out which plugin signals would be required in pretalx. I'd try to provide the authenticator via one class, as we do with [recording providers](https://docs.pretalx.org/en/latest/developer/plugins/recording.html) and [exporters](https://docs.pretalx.org/en/latest/developer/plugins/exporter.html).

Would you be willing to work on things like that? Does that sound like a feasible plan?

**Comment by vmx** [comment]
That sounds great. I don't know much about signals yet, but I guess with the help of the examples above I can work something out.

**Comment by rixx** [comment]
Sure, and I'll be happy to help – but in the end, you can look at the existing examples and either make a similar PR to pretalx, or just tell me about the interface you'd need, and I'll add them for you.

I added the [repository](https://github.com/pretalx/pretalx-socialauth/), put in the files generated by our plugin cookiecutter, and gave you access to it. Feel free to ping me with anything else you need!
